File: examples/c/vector_magnitude.c
/**
 * Vector Magnitude Calculator — C Implementation
 * 
 * Computes the Euclidean norm of an N-dimensional vector.
 * Input: N followed by N double values.
 *
 * DEMO PROGRAM — for testing the RCI Code Equivalence Platform.
 */

#include <stdio.h>
#include <math.h>

#define MAX_DIM 1024

double vector_magnitude(const double *v, int n) {
    double sum = 0.0;
    for (int i = 0; i < n; i++) {
        sum += v[i] * v[i];
    }
    return sqrt(sum);
}

int main(void) {
    int n;
    double v[MAX_DIM];

    scanf("%d", &n);
    if (n <= 0 || n > MAX_DIM) {
        fprintf(stderr, "Error: n must be between 1 and %d\n", MAX_DIM);
        return 1;
    }

    for (int i = 0; i < n; i++) {
        scanf("%lf", &v[i]);
    }

    double mag = vector_magnitude(v, n);
    printf("%.10f\n", mag);

    return 0;
}

File: examples/fortran/vector_magnitude.f90
! Vector Magnitude Calculator — Fortran Implementation
!
! Equivalent to vector_magnitude.c
! NOTE: Arrays are 1-based in Fortran (DO i = 1, n)
!       vs 0-based in C (for i = 0; i < n).
!       The IR comparison layer normalizes this difference.
!
! DEMO PROGRAM — for testing the RCI Code Equivalence Platform.

FUNCTION vector_magnitude(v, n) RESULT(mag)
    IMPLICIT NONE
    INTEGER, INTENT(IN) :: n
    DOUBLE PRECISION, INTENT(IN) :: v(n)
    DOUBLE PRECISION :: mag, sum_sq
    INTEGER :: i

    sum_sq = 0.0D0
    DO i = 1, n
        sum_sq = sum_sq + v(i) * v(i)
    END DO
    mag = DSQRT(sum_sq)
END FUNCTION vector_magnitude

PROGRAM VectorMagnitude
    IMPLICIT NONE
    INTEGER :: n, i
    INTEGER, PARAMETER :: MAX_DIM = 1024
    DOUBLE PRECISION :: v(MAX_DIM), mag
    DOUBLE PRECISION :: vector_magnitude

    READ(*,*) n
    IF (n <= 0 .OR. n > MAX_DIM) THEN
        WRITE(*,*) 'Error: n must be between 1 and', MAX_DIM
        STOP 1
    END IF

    DO i = 1, n
        READ(*,*) v(i)
    END DO

    mag = vector_magnitude(v, n)
    WRITE(*,'(F20.10)') mag

END PROGRAM VectorMagnitude
